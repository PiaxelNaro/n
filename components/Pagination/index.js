// @flow
import React from 'react';
import styled from 'styled-components';
import FaAngleLeft from 'react-icons/lib/fa/angle-left';
import FaAngleRight from 'react-icons/lib/fa/angle-right';
import FlexGrid from '../FlexGrid';

const PAGE_FIRST = 1;

const FlexItem = styled(FlexGrid.Item).attrs({
    flex: 0,
})`
    font-size: 12px;
    white-space: nowrap;
    color: #666;
`;

const PageInput = styled.input`
    width: 30px;
    height: 18px;
    padding: 0;
    text-align: center;
    border: 1px solid #ccc;
    color: #555;
    outline: none;
`;

const NavButton = styled.button.attrs({
    type: 'button',
})`
    border-width: 1px;
    border-style: solid;
    border-color: rgba(0, 0, 0, 0.15) rgba(0, 0, 0, 0.15) rgba(0, 0, 0, 0.25);
    border-radius: 2px;
    background-color: #f6f6f6;
    background-image:  linear-gradient(to bottom, #fff, #e9e9e9);
    box-shadow: inset 0 1px 0 rgba(255, 255, 255, 0.2), 0 1px 2px rgba(0, 0, 0, 0.05);
    color: #999;
    cursor: pointer;
    font-size: 0;
    outline: none;
    padding: 3px 4px;
    &:hover,
    &:focus {
        border: 1px solid #aaa;
        background-color: #ddd;
        background-image: linear-gradient(to bottom, #fff, #ddd);
    }
    &:active {
        box-shadow: inset 0 2px 4px rgba(0, 0, 0, 0.15), 0 1px 2px rgba(0, 0, 0, 0.05);
    }
    &[disabled] {
        background-color: #e6e6e6;
        background-image: none;
        opacity: 0.65;
        box-shadow: none;
        cursor: default;
        &:hover,
        &:focus {
            border-color: rgba(0, 0, 0, 0.15) rgba(0, 0, 0, 0.15) rgba(0, 0, 0, 0.25);
            background-color: #f6f6f6;
            background-image: linear-gradient(to bottom, #ffffff, #e9e9e9);
        }
    }
`;

class Pagination extends React.Component {
    constructor(props: { current: number, onPageChange: number => void, total: number }) {
        super(props);
        this.state = {
            inputValue: props.current,
        };
        this.onInputChange = this.onInputChange.bind(this);
        this.onKeyPress = this.onKeyPress.bind(this);
        this.onPrevClick = this.onPrevClick.bind(this);
        this.onNextClick = this.onNextClick.bind(this);
    }

    props: {
        current: number,
        total: number,
        onPageChange: number => void,
    };

    state: {
        inputValue: string | number,
    };

    onInputChange: Function;
    onKeyPress: Function;
    onPrevClick: Function;
    onNextClick: Function;

    onInputChange(event: SyntheticInputEvent) {
        // remove invalided charater
        this.setState({
            inputValue: event.target.value.replace(/[^0-9]/g, ''),
        });
    }

    onKeyPress(event: SyntheticEvent) {
        if (event.key === 'Enter' && this.state.inputValue) {
            const newInputValue = parseInt(this.state.inputValue, 10);
            if (
                !isNaN(newInputValue) &&
                newInputValue >= PAGE_FIRST &&
                newInputValue <= this.props.total
            ) {
                this.props.onPageChange(newInputValue);
            }
        }
    }

    onPrevClick() {
        if (this.props.current === PAGE_FIRST) {
            return;
        }
        this.props.onPageChange(this.props.current - 1);
    }

    onNextClick() {
        if (this.props.current >= this.props.total) {
            return;
        }
        this.props.onPageChange(this.props.current + 1);
    }

    isInputDisabled(type: 'prev' | 'next') {
        if (type === 'prev') {
            return this.props.current <= PAGE_FIRST;
        }

        return this.props.current >= this.props.total;
    }

    componentWillReceiveProps(nextProps: Object) {
        this.setState({
            inputValue: nextProps.current,
        });
    }

    render() {
        return (
            <FlexGrid inline gutter="5px">
                <FlexItem>
                    Go to page
                </FlexItem>
                <FlexItem>
                    <PageInput
                        type="text"
                        value={this.state.inputValue}
                        onChange={this.onInputChange}
                        onKeyPress={this.onKeyPress}
                    />
                </FlexItem>
                <FlexItem>
                    of
                </FlexItem>
                <FlexItem>
                    {this.props.total}
                </FlexItem>
                <FlexItem>
                    <NavButton
                        onClick={this.onPrevClick}
                        disabled={this.isInputDisabled('prev')}
                    >
                        <FaAngleLeft size={15} />
                    </NavButton>
                </FlexItem>
                <FlexItem>
                    <NavButton
                        onClick={this.onNextClick}
                        disabled={this.isInputDisabled('next')}
                    >
                        <FaAngleRight size={15} />
                    </NavButton>
                </FlexItem>
            </FlexGrid>
        );
    }
}

export default Pagination;
