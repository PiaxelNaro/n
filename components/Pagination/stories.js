// @flow
import React from 'react';
import { storiesOf } from '@kadira/storybook';
import { decorateAction } from '@kadira/storybook-addon-actions';
import Pagination from '.';

const storiesAction = decorateAction([args => args.slice(0, 1)]);
class PaginationDemo extends React.Component {
    handlePageChange: Function;

    constructor() {
        super();
        this.state = {
            current: 2,
            total: 20,
        };
        this.handlePageChange = this.handlePageChange.bind(this);
    }

    handlePageChange(page) {
        this.setState({
            current: page,
        });
    }

    render() {
        return (
            <Pagination
                total={this.state.total}
                current={this.state.current}
                onPageChange={this.handlePageChange}
            />
        );
    }
}

storiesOf('Pagination', module)
    .add('render pagination', () => (
        <Pagination total={10} current={2} onPageChange={storiesAction('onPageChange')} />
    ))
    .add('render dynamic pagination', () => <PaginationDemo />);
