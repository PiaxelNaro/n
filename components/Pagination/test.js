// @flow
import React from 'react';
import { shallow, mount } from 'enzyme';
import Pagination from '.';

describe('<Pagination>', () => {
    it('render children', () => {
        expect(
            shallow(
                <Pagination
                    current={3}
                    total={10}
                    onPageChange={() => {}}
                />
            )
        ).toMatchSnapshot();
    });

    it('render first page, the prev button should be disabled', () => {
        expect(
            shallow(
                <Pagination
                    current={1}
                    total={10}
                    onPageChange={() => {}}
                />
            )
        ).toMatchSnapshot();
    });

    it('render last page, the next button should be disabled', () => {
        expect(
            shallow(
                <Pagination
                    current={10}
                    total={10}
                    onPageChange={() => {}}
                />
            )
        ).toMatchSnapshot();
    });

    it('click prev and next button will invoke callback', () => {
        const onPageChange = jest.fn();
        const pager = mount(
            <Pagination
                current={3}
                total={10}
                onPageChange={onPageChange}
            />
        );
        const buttons = pager.find('button');

        buttons.first().simulate('click');
        expect(onPageChange).toHaveBeenCalledWith(2);
        buttons.last().simulate('click');
        expect(onPageChange).toHaveBeenCalledWith(4);
    });

    it('prev button will be disabled if current is 0', () => {
        const onPageChange = jest.fn();
        const pager = mount(
            <Pagination
                current={1}
                total={10}
                onPageChange={onPageChange}
            />
        );
        pager.find('button').first().prop('onClick')();
        expect(onPageChange).not.toHaveBeenCalled();
    });

    it('next button will be disabled if current is the last', () => {
        const onPageChange = jest.fn();
        const pager = mount(
            <Pagination
                current={10}
                total={10}
                onPageChange={onPageChange}
            />
        );
        pager.find('button').last().prop('onClick')();
        expect(onPageChange).not.toHaveBeenCalled();
    });

    it('test valid input, callback will be invoked', () => {
        const onPageChange = jest.fn();
        const pager = mount(
            <Pagination
                current={3}
                total={10}
                onPageChange={onPageChange}
            />
        );
        const input = pager.find('input').first();
        input.simulate('change', {
            target: {
                value: '10',
            },
        });
        input.simulate('keyPress', { key: 'Test' });
        expect(onPageChange).not.toHaveBeenCalled();

        input.simulate('keyPress', { key: 'Enter' });
        expect(onPageChange).toHaveBeenCalledWith(10);

        // test regexp for input value
        input.simulate('change', {
            target: {
                value: 'abc11d',
            },
        });
        expect(input.prop('value')).toEqual('11');

        // test componentWillReceiveProps(nextProps)
        pager.setProps({
            current: 9,
        });
        expect(input.prop('value')).toEqual(9);
    });

    it('test invalid input, callback will not be invoked', () => {
        const onPageChange = jest.fn();
        const pager = mount(
            <Pagination
                current={3}
                total={10}
                onPageChange={onPageChange}
            />
        );
        const input = pager.find('input').first();
        input.simulate('change', {
            target: {
                value: '99',
            },
        });
        input.simulate('keyPress', { key: 'Enter' });
        expect(onPageChange).not.toHaveBeenCalled();
    });
});
