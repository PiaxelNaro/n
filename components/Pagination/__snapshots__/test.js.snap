// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination> render children 1`] = `
<FlexGrid
  gutter="5px"
  inline={true}
>
  <Pagination__FlexItem>
    Go to page
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__PageInput
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value={3}
    />
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    of
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    10
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__NavButton
      disabled={false}
      onClick={[Function]}
    >
      <FaAngleLeft
        size={15}
      />
    </Pagination__NavButton>
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__NavButton
      disabled={false}
      onClick={[Function]}
    >
      <FaAngleRight
        size={15}
      />
    </Pagination__NavButton>
  </Pagination__FlexItem>
</FlexGrid>
`;

exports[`<Pagination> render first page, the prev button should be disabled 1`] = `
<FlexGrid
  gutter="5px"
  inline={true}
>
  <Pagination__FlexItem>
    Go to page
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__PageInput
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value={1}
    />
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    of
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    10
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__NavButton
      disabled={true}
      onClick={[Function]}
    >
      <FaAngleLeft
        size={15}
      />
    </Pagination__NavButton>
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__NavButton
      disabled={false}
      onClick={[Function]}
    >
      <FaAngleRight
        size={15}
      />
    </Pagination__NavButton>
  </Pagination__FlexItem>
</FlexGrid>
`;

exports[`<Pagination> render last page, the next button should be disabled 1`] = `
<FlexGrid
  gutter="5px"
  inline={true}
>
  <Pagination__FlexItem>
    Go to page
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__PageInput
      onChange={[Function]}
      onKeyPress={[Function]}
      type="text"
      value={10}
    />
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    of
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    10
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__NavButton
      disabled={false}
      onClick={[Function]}
    >
      <FaAngleLeft
        size={15}
      />
    </Pagination__NavButton>
  </Pagination__FlexItem>
  <Pagination__FlexItem>
    <Pagination__NavButton
      disabled={true}
      onClick={[Function]}
    >
      <FaAngleRight
        size={15}
      />
    </Pagination__NavButton>
  </Pagination__FlexItem>
</FlexGrid>
`;
